homotopy-model/src/proof.rs

```diff
@@ -728,22 +725,41 @@ impl ProofState {
     }
 
     /// Handler for [Action::Suspend].
-    fn suspend(&mut self, reduced: bool) -> bool {
+    fn suspend(&mut self, kind: SuspensionKind) -> bool {
         use homotopy_common::tree::Node;
 
+        let mut kind = kind;
         let mut new_signature: Signature = Default::default();
-
-        // New generators need to be fresh
         let id = self.signature.next_generator_id();
-        let source = Generator::new(id, 0);
-        let target = if reduced {
-            new_signature.insert(source, Diagram0::from(source), "Base", false);
-            source
-        } else {
-            new_signature.insert(source, Diagram0::from(source), "Base Source", false);
-            let target = Generator::new(id + 1, 0);
-            new_signature.insert(target, Diagram0::from(target), "Base Target", false);
-            target
+        let genz_count = self
+            .signature
+            .generator_iter()
+            .filter(|g| g.dimension == 0)
+            .count();
+        let (source, target) = match (genz_count, kind) {
+            // If signature is empty quit
+            (0, _) => {
+                return false;
+            }
+            // If there is not a unique 0-cell we perform a reduced suspension
+            (1, SuspensionKind::Abelian) => {
+                let base = *self.signature.generator_iter().next().unwrap();
```

Comment:
This might not be dimension 0 (if you reordered the generators) which causes a panic.

Reply:
Thanks for finding this! Fixed